File: Cooking/domain/src/main/java/com/thangcao/cooking/common/base/usecase/UseCase.java
package com.thangcao.cooking.common.base.usecase;

import com.thangcao.cooking.common.base.output.OutputObserver;

/**
 * @author ThangCao
 */
public abstract class UseCase<Input, Output> {

    public void execute(Input param, OutputObserver<Output> baseOutput) {
        try {
            baseOutput.onNext(handleBusiness(param));
            baseOutput.onComplete();
        } catch (Exception ex) {
            baseOutput.onError(ex);
        }
    }
    protected abstract Output handleBusiness(Input param);

}

File: Cooking-Backend/src/shared/core/UseCase.ts
import { IOutput } from "../core/Output";

export default abstract class UseCase<Input, Output> {

    output: IOutput<Output>;
    abstract buildUseCase(input?: Input): Promise<Output>;
    
    execute(input?: Input): Promise<Output> {
        return this.buildUseCase(input);
    }
}